### PR TITLE
Bug/csi 3038 node pods in status creating due to bad secrets corelation

### PR DESCRIPTION
**Pull request by matancarmeli7**

### CODE REVIEW

File: pkg/internal/ibmblockcsi/ibmblockcsi.go [hunk 1]
```diff
@@ -67,7 +68,7 @@ func (c *IBMBlockCSI) GetLabels() labels.Set {
 }
 
 // GetAnnotations returns all the annotations to be set on all resources
-func (c *IBMBlockCSI) GetAnnotations() labels.Set {
+func (c *IBMBlockCSI) GetAnnotations(ds_restarted_key string , ds_restarted_value string) labels.Set {
```
Comment (zingero) [inline pkg/internal/ibmblockcsi/ibmblockcsi.go:71]: Functions should do one and only one thing.
In this case, the function should return the annotations.
Adding the key and value makes the function do two things.
Also, usually `get` functions don't change the data structure (`set functions do)

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 2]
```diff
@@ -505,6 +525,58 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+		Namespace: instance.Namespace,
+	}, node)
+
+	if err != nil {
+		return err, node
+	}
+	return err, node
+}
+
+func (r *ReconcileIBMBlockCSI) getDriverPodsStatus(instance *ibmblockcsi.IBMBlockCSI,
```
Comment (zingero) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:564]: What is the reason for returning two values?
I suggest breaking this function into two functions (one for the controller and one for the driver)
Then, the local variables will be redundant.

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 3]
```diff
@@ -505,6 +525,58 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+	node := &appsv1.DaemonSet{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{
+		Name:      oconfig.GetNameForResource(oconfig.CSINode, instance.Name),
+		Namespace: instance.Namespace,
+	}, node)
+
+	if err != nil {
+		return err, node
+	}
+	return err, node
```
Comment (zingero) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:561]: Why do we need the `if` statements?

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 4]
```diff
@@ -505,6 +525,58 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+			controllerStatefulset := &appsv1.StatefulSet{}
+			err := r.client.Get(context.TODO(), types.NamespacedName{
+				Name:      oconfig.GetNameForResource(oconfig.CSIController, instance.Name),
+				Namespace: instance.Namespace,
+			}, controllerStatefulset)
+		
+			if err != nil {
+				return err, controllerStatefulset
+			}
+			return err, controllerStatefulset
```
Comment (zingero) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:548]: Why do we need the if statements?

Comment (zingero) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:548]: Also, usually, when we return the error and another value, the error is the second value.

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 3 above)
Comment (zingero) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:561]: Also, usually, when we return the error and another value, the error is the second value.

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 5]
```diff
@@ -56,7 +56,14 @@ import (
...
 // ReconcileTime is the delay between reconciliations
-const ReconcileTime = 30 * time.Second
+const (
+	ReconcileTime = 30 * time.Second
+	Node = "node"
+	Controller = "controller"
+)
+
+var ds_restarted_key = ""
+var ds_restarted_value = ""
```
Comment (zingero) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:66]: What is `ds`?

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 6]
```diff
@@ -320,60 +328,46 @@ func (r *ReconcileIBMBlockCSI) getAccessorAndFinalizerName(instance *ibmblockcsi
...
-	instance.Status.ControllerReady = controllerStatefulset.Status.ReadyReplicas == controllerStatefulset.Status.Replicas
-	instance.Status.NodeReady = node.Status.DesiredNumberScheduled == node.Status.NumberAvailable
+	if !instance.Status.ControllerReady {
+		err := r.getControllerPod(controllerStatefulset, controllerPod)
+		if err != nil {
+			logger.Error(err, "failed to get controller pod")
+			return err
+		}
+
+		for _, containerstatus := range controllerPod.Status.ContainerStatuses {
```
Comment (zingero) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:348]: camelCase:
`containerStatus`

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 7]
```diff
@@ -320,60 +328,46 @@ func (r *ReconcileIBMBlockCSI) getAccessorAndFinalizerName(instance *ibmblockcsi
...
+					logger.Info("controller requires restart",
+											"ReadyReplicas", controllerStatefulset.Status.ReadyReplicas,
+											"Replicas", controllerStatefulset.Status.Replicas)
+					r.restartControllerPod(controllerPod)
+				}
+			}
+		}
+	}
+
+	instance.Status.ControllerReady, instance.Status.NodeReady = r.getDriverPodsStatus(instance,
```
Comment (zingero) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:360]: Why do we need to assign the values in the `instance.Status`?
The function already does that.
I see two options:
1. remove the assignment in the function
2. remove the assignment here and rename the function since it doesn't get anything.

**Review by zingero — CHANGES_REQUESTED** [review]
Run the seamless upgrade job, please.

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 5 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:66]: daemonSet

Comment (zingero) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:66]: ohhh
I thought it's "data structure".
Please rename it accordingly.

File: pkg/internal/ibmblockcsi/ibmblockcsi.go (hunk 1 above)
Comment (zingero) [reply pkg/internal/ibmblockcsi/ibmblockcsi.go:71]: I know it returns one thing.
It does more than one thing though.
For more information:
https://stackoverflow.com/questions/475675/when-is-a-function-too-long

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 8]
```diff
@@ -535,46 +534,38 @@ func (r *ReconcileIBMBlockCSI) getRestartedAtAnnotation(Annotations map[string]s
...
+	return node, err
 }
 
-func (r *ReconcileIBMBlockCSI) getDriverPodsStatus(instance *ibmblockcsi.IBMBlockCSI, 
-	controller *appsv1.StatefulSet, node *appsv1.DaemonSet) (bool, bool) {
+func (r *ReconcileIBMBlockCSI) getControllerPodsStatus(instance *ibmblockcsi.IBMBlockCSI, controller *appsv1.StatefulSet) bool {
 	ControllerReady := false
+
+	ControllerReady = controller.Status.ReadyReplicas == controller.Status.Replicas
+	return ControllerReady
```
Comment (zingero) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: Why do we need the local variable? 
```suggestion
	return controller.Status.ReadyReplicas == controller.Status.Replicas 
```

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 9]
```diff
@@ -535,46 +534,38 @@ func (r *ReconcileIBMBlockCSI) getRestartedAtAnnotation(Annotations map[string]s
...
+}
+
+func (r *ReconcileIBMBlockCSI) getNodePodsStatus(instance *ibmblockcsi.IBMBlockCSI, node *appsv1.DaemonSet) bool {
 	NodeReady := false
```
Comment (zingero) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:565]: Same

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 8 above)
Comment (zingero) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: Also, the function name says that it returns the status, but actually, we return something else.
Rename (also for the node function)

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 10]
```diff
@@ -355,25 +350,23 @@ func (r *ReconcileIBMBlockCSI) updateStatus(instance *ibmblockcsi.IBMBlockCSI, o
...
 			}
-			if originalStatus.ControllerReady || !r.areAllPodImagesSynced(controllerStatefulset, controllerPod) {
-				logger.Info("controller requires restart",
-							"ReadyReplicas", controllerStatefulset.Status.ReadyReplicas,
-							"Replicas", controllerStatefulset.Status.Replicas)
-				controllerRestart = true
+	
+			for _, containerStatus := range controllerPod.Status.ContainerStatuses {
+				if containerStatus.State.Waiting != nil {
+					if containerStatus.State.Waiting.Reason == "ImagePullBackOff" {
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:356]: why is this needed?
I don't think we should restart the pod for each faulty container

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 11]
```diff
@@ -505,6 +524,44 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+	return node, err
+}
+
+func (r *ReconcileIBMBlockCSI) doesControllerPodsRunning(instance *ibmblockcsi.IBMBlockCSI, controller *appsv1.StatefulSet) bool {
+	return controller.Status.ReadyReplicas == controller.Status.Replicas
+}
+
+func (r *ReconcileIBMBlockCSI) doesNodePodsRunning(instance *ibmblockcsi.IBMBlockCSI, node *appsv1.DaemonSet) bool {
+	return node.Status.DesiredNumberScheduled == node.Status.NumberAvailable
+}
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:563]: why do they need the `instance`?
boolean functions usually start with `is`, and in this case the high-level question is `is<Component>Ready`

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 12]
```diff
@@ -505,6 +524,44 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+
+func (r *ReconcileIBMBlockCSI) getNodeK8sObject(instance *ibmblockcsi.IBMBlockCSI) (*appsv1.DaemonSet, error) {
+	node := &appsv1.DaemonSet{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{
+		Name:      oconfig.GetNameForResource(oconfig.CSINode, instance.Name),
+		Namespace: instance.Namespace,
+	}, node)
+
+	return node, err
+}
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:555]: rename `k8sObject` to the specific k8s object, and fix over indentation

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 10 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:356]: we still need to handle the problem with statefulsets that doesn't redoply pods in ImagePullBackOff even though the user updates their images

Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:356]: oh, then you can still use `areAllPodImagesSynced` for that

Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:356]: what is `areAllPodImagesSynced`? this is an existing function? or its a function that you want me to add?

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 12 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 11 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 10 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/internal/ibmblockcsi/ibmblockcsi.go (hunk 1 above)
Comment (zingero) [reply pkg/internal/ibmblockcsi/ibmblockcsi.go]: The function changes the DS. According to its name, it should be read-only (const function in C++)

Comment (matancarmeli7) [reply pkg/internal/ibmblockcsi/ibmblockcsi.go]: I don't think that it changes the DS. Its only returns the wanted annotations that should be on the DS/STS

Comment (zingero) [reply pkg/internal/ibmblockcsi/ibmblockcsi.go]: I meant DS == Data Structure, no Daemon Set.
Sorry for the confusion.

Comment (matancarmeli7) [reply pkg/internal/ibmblockcsi/ibmblockcsi.go]: done

**Review by zingero — APPROVED** [review]

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 10 above)
Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: I still don't see why we should restart the pod for each faulty container, not why we need to check the waiting reason.
can't we just keep the original code here (minus the `originalStatus` check)?

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 12 above)
Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: no need for the word `Object`

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 11 above)
Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: could you please renaming the functions to `isControllerRead` and `isNodeReady` respectively?

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 13]
```diff
@@ -355,25 +350,23 @@ func (r *ReconcileIBMBlockCSI) updateStatus(instance *ibmblockcsi.IBMBlockCSI, o
...
-			logger.Info("node rollout requires restart",
-				"DesiredNumberScheduled", node.Status.DesiredNumberScheduled,
-				"NumberAvailable", node.Status.NumberAvailable)
-			nodeRolloutRestart = true
-		}
 		phase = csiv1.DriverPhaseCreating
 	}
 	instance.Status.Phase = phase
 	instance.Status.Version = oversion.DriverVersion
+	logger.Info("updating IBMBlockCSI status")
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:369]: I don't think this is true (it's true only inside the following `if`)

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 10 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: We wanted to remove the original code and move it to the service account reconciler.
The original code was also restarting the controller pods for each faulty container, this code is restarting only when there is an ImagePullBackOff and the function `areAllPodImagesSynced` returns false so it minimizes the restating action to less events

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 14]
```diff
@@ -493,6 +468,50 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+				
+				if rErr != nil {
+					return rErr
+				}
+			}
+			if strings.Contains(sa.Name, Node) {
+				nodelogger := log.WithValues("Resource Type", "Node DaemonSet")
+				nodelogger.Info("node rollout requires restart",
+				"DesiredNumberScheduled", nodeDaemonSet.Status.DesiredNumberScheduled,				
+				"NumberAvailable", nodeDaemonSet.Status.NumberAvailable)
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:505]: please indent function parameters wherever applicable, to align to current project convention

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 13 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:369]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 12 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 10 above)
Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: the original code did **not** restart the controller pod **for each** faulty container, it restarted the pod **once** if **any** container image was not in sync with the statefulset.
so current diff can potentially **add** restart events

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 11 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 14 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:505]: do you have an example?

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 15]
```diff
@@ -493,6 +468,50 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+			if err != nil {
+				return err
+			}
+
+			nodeDaemonSet, err := r.getNodeDaemonSetObject(instance)
+			if err != nil {
+				return err
+			}
+
+			if strings.Contains(sa.Name, Controller) {
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:481]: please compare between full strings (i.e. without `Contains`).
perhaps with [`GetNameForResource`](https://github.com/IBM/ibm-block-csi-operator/blob/9a30d70922dbfa46429c8ce6df9c16684510b76d/pkg/controller/ibmblockcsi/syncer/csi_controller.go#L109)

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 10 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: but the current diff also restart the pod once if any container image was not in sync with the statefulset, but when there is `ImagePullBackOff`, I can remove the if of the `ImagePullBackOff` if you like

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 14 above)
Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:504]: e.g. [here](https://github.com/IBM/ibm-block-csi-operator/blob/9a30d70922dbfa46429c8ce6df9c16684510b76d/pkg/controller/ibmblockcsi/ibmblockcsi_controller.go#L420) (though it should be even a bit more indented, to be aligned with the 1st argument)

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 10 above)
Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: it seems to me as if the current diff restarts **for each** faulty container.
what am I missing here?

Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: I would remove the `ImagePullBackOff` regardless

Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: I will do it like you suggested at the beginning

Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 14 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 16]
```diff
@@ -56,7 +56,14 @@ import (
...
 
 // ReconcileTime is the delay between reconciliations
-const ReconcileTime = 30 * time.Second
+const (
+	ReconcileTime = 30 * time.Second
+	Node = "node"
+	Controller = "controller"
+)
+
+var daemonSet_restarted_key = ""
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:65]: this can be a `fmt.Sprintf("%s/restartedAt", oconfig.APIGroup)` "const" here

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 17]
```diff
@@ -56,7 +56,14 @@ import (
...
+var daemonSet_restarted_value = ""
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:66]: use camelCase everywhere

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 15 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 18]
```diff
@@ -56,7 +56,12 @@ import (
 )
 
 // ReconcileTime is the delay between reconciliations
-const ReconcileTime = 30 * time.Second
+const (
+	ReconcileTime = 30 * time.Second
+)
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:61]: no need for round brackets (parenthesis)

Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 17 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 19]
```diff
@@ -493,6 +460,50 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+			if err != nil {
+				return err
+			}
+
+			nodeDaemonSet, err := r.getNodeDaemonSet(instance)
+			if err != nil {
+				return err
+			}
+
+			if config.GetNameForResource(config.CSIControllerServiceAccount, instance.Name) == sa.Name {
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:474]: please give this return value a name, before comparing it.
e.g. `controllerServiceAccountName` (same for node)

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 16 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: I didn't do it because I wanted that as long as the `restartedAt` annotation won't be on the `daemonset` it will stay empty, that the sycer will know that there isn't a new annotation that it needs to take care of

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 19 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 16 above)
Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: it can check the value, I don't think the key is interesting

Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: oh, you don't want `reconcileServiceAccount` to return anything...

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 20]
```diff
@@ -505,6 +517,44 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+}
+
+func (r *ReconcileIBMBlockCSI) getControllerStatefulSet(instance *ibmblockcsi.IBMBlockCSI) (*appsv1.StatefulSet, error) {
+			controllerStatefulset := &appsv1.StatefulSet{}
+			err := r.client.Get(context.TODO(), types.NamespacedName{
+				Name:      oconfig.GetNameForResource(oconfig.CSIController, instance.Name),
+				Namespace: instance.Namespace,
+			}, controllerStatefulset)
+
+			return controllerStatefulset, err
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:537]: please align all indentations to current project conventions.
e.g. this seems over-indented to me.
I think Ctrl+Alt+L

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 16 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: `reconcileServiceAccount` return only error, because `daemonSetRestartedKey` and `daemonSetRestartedValue` are global variables

Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: so I can check the value and move the to the const section if you like

Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: I will try to find a better way

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 21]
```diff
@@ -493,6 +461,50 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+				if rErr != nil {
+					return rErr
+				}
+			}
+			if nodeServiceAccountName == sa.Name {
+				nodelogger := log.WithValues("Resource Type", "Node DaemonSet")
+				nodelogger.Info("node rollout requires restart",
+								"DesiredNumberScheduled", nodeDaemonSet.Status.DesiredNumberScheduled,				
+								"NumberAvailable", nodeDaemonSet.Status.NumberAvailable)
+				nodelogger.Info("csi node stopped being ready - restarting it")
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:499]: why the separate loggers? consider using this function's common logger

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 22]
```diff
@@ -493,6 +461,50 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+			nodeDaemonSet, err := r.getNodeDaemonSet(instance)
+			if err != nil {
+				return err
+			}
+
+			if controllerServiceAccountName == sa.Name {
+				controllerlogger := log.WithValues("Resource Type", "Controller")
+				controllerPod := &corev1.Pod{}
+				err := r.getControllerPod(controllerStatefulset, controllerPod)
+				if err != nil {
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:479]: what about `errors.IsNotFound(err)`? what if the driver just wasn't deployed yet?

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 21 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:499]: because the resource is different and in the logs I don't see `node rollout requires restart, resource service account` for example

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 22 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 16 above)
Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: could you open a ticket for it and write its number here?
I guess it won't be for this release

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 23]
```diff
@@ -493,6 +464,52 @@ func (r *ReconcileIBMBlockCSI) reconcileServiceAccount(instance *ibmblockcsi.IBM
...
+			if err != nil {
+				return err
+			}
+
+			if controllerServiceAccountName == sa.Name {
+				controllerlogger := log.WithValues("Resource Type", "Controller")
+				controllerPod := &corev1.Pod{}
+				err := r.getControllerPod(controllerStatefulset, controllerPod)
+				if errors.IsNotFound(err) {
+					return nil
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:483]: will this prevent from the second service account to be reconciled?

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 20 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 23 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:483]: I think maybe I will add the `if errors.IsNotFound(err)` in the `getControllerPod`, and remove it from checking here and in the other place that we call to `getControllerPod` function. What do you think?

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 21 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 16 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 23 above)
Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:483]: it would still require every caller to check whether the pod is not nil, but I guess it's better than nothing?
I think here specifically the caller doesn't need the pod, so it would be better if it could call a function that doesn't require the pod as a parameter

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 24]
```diff
@@ -58,6 +58,10 @@ import (
 // ReconcileTime is the delay between reconciliations
 const ReconcileTime = 30 * time.Second
 
+// ticket to redundant those vars - CSI-3071
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:61]: to remove*

Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:61]: done

Comment (oriyarde) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:61]: I don't see it

Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: now it should be applied, I didn't write the done as part of the review

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 23 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go [hunk 25]
```diff
@@ -424,7 +388,36 @@ func (r *ReconcileIBMBlockCSI) areAllPodImagesSynced(controllerStatefulset *apps
...
+		return nil
+	} else if err != nil {
+		logger.Error(err, "failed to get controller pod")
+		return err
+	}
+
+	return r.restartControllerPodfromStatefulSet(logger, instance, controllerStatefulset, controllerPod)
+}
+
+func (r *ReconcileIBMBlockCSI) restartControllerPodfromStatefulSet(logger *logf.DelegatingLogger, instance *ibmblockcsi.IBMBlockCSI,
```
Comment (oriyarde) [inline pkg/controller/ibmblockcsi/ibmblockcsi_controller.go:414]: does it really need the `instance`?

**Review by oriyarde — APPROVED** [review]
nice job!

File: pkg/controller/ibmblockcsi/ibmblockcsi_controller.go (hunk 25 above)
Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: no, I will remove it

Comment (matancarmeli7) [reply pkg/controller/ibmblockcsi/ibmblockcsi_controller.go]: done
